### PR TITLE
Move `react` to peerDependencies

### PR DESCRIPTION
**Pull request by skevy**
This PR moves `react` from dependencies to peerDependencies.

In general, this would have only been important for those people using packages that depend on `react` and were using npm@2...npm@3 would automatically de-dupe.

However, when #5812 gets merged, dependencies will be scoped to react-native (on both npm@2 & npm@3), thus breaking projects that are using a package like `react-redux` for example, which depends on `react`. There would be two copies of React installed, and due to the use of haste modules in `react`, this would break the packager and cause naming collisions.

This PR does three things -

1. Moves the dependency from dependencies to peerDependencies
2. Updates the local-cli to run `npm install react --save` when a new project is initialized.
3. Updates `react-native upgrade` to warn if `react` is not listed in the package.json's dependencies.

**Note: This will require a shrinkwrap update.**

### CODE REVIEW

**Comment by skevy** [comment]
cc @ide

File: local-cli/upgrade/upgrade.js [hunk 1]
```diff
@@ -44,6 +44,19 @@ module.exports = function upgrade(args, config) {
               'https://github.com/facebook/react-native/releases/tag/v' + semver.major(v) + '.' + semver.minor(v) + '.0'
             )
           );
+
+          // >= v0.21.0, we require react to be a peer depdendency
+          if (semver.gte(v, '0.21.0') && !pak.dependencies['react']) {
+            console.log(
+              chalk.yellow(
+                '\nYour \'package.json\' file doesn\'t seem to have \'react\' as a dependency.\n' +
```
Comment (ide) [inline local-cli/upgrade/upgrade.js]: style suggestion: I often use template strings for messages so quotes don't have to be escaped

File: react-native-cli/index.js [hunk 2]
```diff
@@ -188,11 +188,11 @@ function createProject(name, verbose) {
 }
 
 function run(root, projectName) {
-  exec('npm install --save react-native', function(e, stdout, stderr) {
+  exec('npm install --save react react-native', function(e, stdout, stderr) {
```
Comment (frantic) [inline react-native-cli/index.js]: might be worth moving `npm install --save react` to `local-cli`? This way existing users won't have to upgrade global `react-native-cli`

Comment (skevy) [inline react-native-cli/index.js]: This is a great idea. I'll figure out a good way to do this.

File: local-cli/upgrade/upgrade.js (hunk 1 above)
Comment (skevy) [inline local-cli/upgrade/upgrade.js]: @ide I was just matching other places in the file. I guess I could update the whole file...are template strings supported everywhere now? (the local-cli isn't run through babel-register)

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-native",
-  "version": "0.0.0-master",
+  "version": "0.22.0",
```
Comment (mkonicek) [inline package.json]: I think we want to keep the version here to make it clear we're on master?

**Comment by mkonicek** [comment]
> Updates react-native-cli to install react ... this will require a release of a new version of rn CLI

Is this still relevant? The CLI is here: https://github.com/facebook/react-native/blob/master/react-native-cli/index.js

What happens when people don't run `react-native upgrade` and simply bump the version of `react-native` in the package.json of their apps?

File: local-cli/generator/index.js [hunk 4]
```diff
@@ -82,5 +82,13 @@ module.exports = yeoman.generators.NamedBase.extend({
...
       );
     }
+  },
+
+  install: function() {
+    if (this.options.upgrade) {
+      return;
+    }
+
+    this.npmInstall('react', { '--save': true });
```
Comment (mkonicek) [inline local-cli/generator/index.js]: Does this respect the declared version of React via `peerDependencies` (0.14.5)?

**Comment by mkonicek** [comment]
I'm not the best person to review this. cc @martinbigio

File: package.json (hunk 3 above)
Comment (skevy) [inline package.json]: This was a remnant of testing in the middle of the night and me being tired. Will fix. :)

**Comment by skevy** [comment]
@mkonicek that note is not still relevant...I took @frantic's advice and moved the code that does anything re: installing react-native to local-cli. PR description has been updated accordingly.

File: local-cli/upgrade/upgrade.js (hunk 1 above)
Comment (ide) [inline local-cli/upgrade/upgrade.js]: IIRC they shipped with Node 4

File: local-cli/generator/index.js (hunk 4 above)
Comment (frantic) [inline local-cli/generator/index.js]: very good question, might be worth installing the version we know matches react-native's one, we can get it from RN's package.json's peerDependencies.

**Comment by mkonicek** [comment]
Cool, thanks @skevy! I'll let @martinbigio or @ide merge this as I'm not that familiar with npm.

**Comment by ide** [comment]
I restarted the E2E tests since the failure looked related to the npm repo. Let's get the E2E tests passing and then merge this. We have a bunch of time including a 2-week RC period to shake out issues.

File: local-cli/generator/index.js (hunk 4 above)
Comment (ide) [inline local-cli/generator/index.js]: +1 to reading the version from RN package.json

**Comment by ide** [comment]
E2E tests look good. I took a closer look at the Obj-C tests and found that they run the UIExplorer. So @skevy I think we either need to add "npm install react" to travis.yml (hacky solution but has some nice properties, like we install all of RN's dependencies just once) or set up a package.json in each example project and be sure to run something like `cd Examples/UIExplorer && npm install` in the Obj-C test

**Comment by skevy** [comment]
@ide I don't think the obj-c test failed because of this change.

It's still installing `react`:

asdf

That's because npm only doesn't install peer deps automatically when it's installing dependencies. When you run `npm install` in the root of a project, like React Native, it will install dev/prod/peer deps.

**Comment by ide** [comment]
Snazzy. move fast break things, let's do it

**Comment by ide** [comment]
@facebook-github-bot shipit

**Comment by skevy** [comment]
Excuse me @ide I think it's "move fast with STABLE INFRA".

Get it right.

**Comment by skevy** [comment]
In all seriousness though, somebody (@mkonicek?) stop this import. I didn't fix the npm install command in the generator, and would like to do that before this goes to core.

**Comment by skevy** [comment]
Eh. nvm. I'll just fix in a follow up diff. This works for now.

**Comment by ide** [comment]
This will get Relay working. We can fix things like getting the react version right shortly after.

**Comment by skevy** [comment]
*almost will get Relay working...still waiting on #5084 :(

**Comment by mkonicek** [comment]
> Note: This will require a shrinkwrap update.

Can you elaborate please? @davidaurelio do we need to do anything internally?

> Snazzy. move fast break things, let's do it

@skevy Are you happy with the PR now? Is there a followup we should merge?

**Comment by mkonicek** [comment]
@skevy I'm trying to understand the note about the shrinkwrap file - why not update it as part of this PR? https://github.com/facebook/react-native/blob/master/npm-shrinkwrap.json

**Comment by skevy** [comment]
@mkonicek honestly I didn't update the shrinkwrap because I thought that you guys had a process internally that would update the shrinkwrap automatically when deps were changed. I'm not sure where I heard that :)

If you don't mind updating it for this change (given that this is already merged in)...and then in the future any PR's I send that change dependencies, I'll make sure I update shrinkwrap.

**Comment by ide** [comment]
The issue with updating shrinkwrap.json (as I understand it) is that FB might not have the correct versions of the modules in trunk, so the shrinkwrap had to be generated by FB. 

This might be addressed recently by the FB npm package tools. Ideally what would happen is PRs would include npm-shrinkwrap.json, then something inside of FB would run `npm install` and get all of the packages, and then feed the packages into FB's npm tooling. That way npm-shrinkwrap.json acts as a source of truth.

**Comment by davidaurelio** [comment]
This breaks our internal shrinkwrap process @mkonicek @bestander @martinbigio

**Comment by mkonicek** [comment]
@skevy I just created a new app with npm2. `package.json` has deps as follows:

    "dependencies": {
      "react": "^0.14.7",
      "react-native": "^0.22.0"
    }

I'm using npm2. When starting the packager using `npm start` I get:

    Failed to build DependencyGraph: @providesModule naming collision:
      Duplicate module name: ViewportMetrics
     Paths: /private/tmp/Zero22/node_modules/react/lib/ViewportMetrics.js collides with /private/tmp/Zero22/node_modules/react-native/node_modules/react/lib/ViewportMetrics.js

So it looks like this PR fixed collisions for some use cases but broke them for others (specifically, no one is able to work on their React Native apps if they're using npm 2).

**Comment by mkonicek** [comment]
@skevy tells me this is resolved by https://github.com/facebook/node-haste/pull/33

**Pull request closed without merging**